clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -7430,6 +7430,169 @@ static void handleSYCLIntelFPGAMaxConcurrencyAttr(Sema &S, Decl *D,
   S.AddSYCLIntelFPGAMaxConcurrencyAttr(D, A, E);
 }
 
+// Returns true if an error occured.
+static bool evaluateAddIRAttributesArgs(Expr **ArgsBegin, size_t ArgsSize,
```

Comment:
This function is a bit big. I think splitting it based on diagnostic checks i.e. calling smaller functions from this function for various diagnostic checks will make it more readable.

Reply:
It seems `evaluateAddIRAttributesArgs` was due for a bit of a cleanup. I have applied your other suggestions and have moved some of the functionality of it into separate functions. Let me know what you think.

Reply:
> Can we rename this as `checkAddIRAttributesArgs` for consistency?

In addition to checking the arguments, it also evaluates them, whereas the other auxiliary functions only do the checks. Maybe a more fitting name would be `evaluateAndCheckAddIRAttributesArgs` (or `checkAndEvaluate...`) but I am not convinced it is much better than the current name.

> Nit: Why is the argument `ArgsBegin` and not just `Args`. I was confused by code in some parts below because I kept thinking this variable held first argument

I think it was just inherited from `Attr->args_begin()` in the argument at the callsites. I will change it, no problem.

Reply:
> > Can we rename this as `checkAddIRAttributesArgs` for consistency?
> 
> In addition to checking the arguments, it also evaluates them, whereas the other auxiliary functions only do the checks. Maybe a more fitting name would be `evaluateAndCheckAddIRAttributesArgs` (or `checkAndEvaluate...`) but I am not convinced it is much better than the current name.

Oh right. You can leave it as is then. Thanks for clarifying!

Reply:
`ArgsBegin` has been renamed to `Args`.